fontconfig: 2.12.6 -> 2.13.92

worldofpeace opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
~This currently fails to build at `postInstall` with~
```
/nix/store/hmqnlh2nvjc4893bxyr0n9i3zyvsh5c3-fontconfig-2.13.92/share/xml/fontconfig/fonts.dtd:127: parser error : xmlParseElementDecl: 'EMPTY', 'ANY' or '(' expected
<!ELEMENT reset-dirs >
                     ^
/nix/store/hmqnlh2nvjc4893bxyr0n9i3zyvsh5c3-fontconfig-2.13.92/share/xml/fontconfig/fonts.dtd:127: parser error : Content error in the external subset
<!ELEMENT reset-dirs >
                     ^
unable to parse /nix/store/hmqnlh2nvjc4893bxyr0n9i3zyvsh5c3-fontconfig-2.13.92/etc/fonts/fonts.conf
```

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @


Closes: https://github.com/NixOS/nixpkgs/issues/73725

worldofpeace committed:
fontconfig: 2.12.6 -> 2.13.92

worldofpeace commented:
Is the reset-dirs element invalid ?
* https://gitlab.freedesktop.org/fontconfig/fontconfig/commit/def1d00036a4e828382027292a167203c6c7a0b4

jtojnar commented:
The DTD looks wrong. The element should have `EMPTY` as a content specifier.

jtojnar commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -18,28 +28,47 @@ let
-  buildInputs = [ expat ];
+  nativeBuildInputs = [
+    gperf
+    libuuid
```
Why is this needed?

worldofpeace replied:
previous pr had it and I recall it failing without it. can try without it locally though.

jtojnar replied:
I would expect it go to `buildInputs`. But the package has a ton of stuff in it so maybe it is indeed used during build (it is actually alias of util-linux here).

worldofpeace replied:
seems to be fine so I removed it.

jtojnar commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -18,28 +28,47 @@ let
+    libuuid
+    libxslt
+    pkgconfig
+    autoreconfHook # drop with fcada522913e5e07efa6367eff87ace9f06d24c8 patch.
```
Is not this only necessary to build the test?

worldofpeace replied:
Yeah, since the patch touches a makefile I believe we have to run it.

jtojnar replied:
Well, I would expect that witout the hook, the Makefile change would just be ignored. Since it is just about building an extra test, we do not really need it.

worldofpeace replied:
see changes in force push :+1:

worldofpeace commented:
Un-drafted because it isn't broke anymore. Thanks @jtojnar for the quick patch :sparkles:

dtzWill commented:
Does `configVersion` need to be changed, and what does this mean for `fontconfig_210`? Do we need three nested sets of conf directories now, or only the two as we have presently?
(and how do we know/determine the answers to these questions?)

The above matters for having a good story for folks using paths built against current fontconfig, etc...

worldofpeace commented:
> Does `configVersion` need to be changed, and what does this mean for `fontconfig_210`? Do we need three nested sets of conf directories now, or only the two as we have presently?
> (and how do we know/determine the answers to these questions?)
> 
> The above matters for having a good story for folks using paths built against current fontconfig, etc...

Probably all good questions for its maintainer.
I don't think I can answer any of those questions.
Thanks for considering backwards compat though, but we do really need to upgrade our fontconfig.

jtojnar commented:
What is wrong with our current version? Is not `x.13` unstable? `x.y.92`, I would definitely expect to be.

worldofpeace commented:
> What is wrong with our current version? Is not `x.13` unstable? `x.y.92`, I would definitely expect to be.

Yes it is a development release, but I'd really like to get it tested while still very early in 20.03 dev.

worldofpeace commented:
@jtojnar Though stable is actually 2.13.1

sauyon commented:
I think fontconfig made breaking configuration changes in 2.13, so maybe `configVersion` should be bumped?

worldofpeace committed:
fontconfig: bump configVersion to 2.13

worldofpeace commented:
> I think fontconfig made breaking configuration changes in 2.13, so maybe `configVersion` should be bumped?

I guess that would be adding a new element. However, I'm not certain how this works in the first place in nixpkgs. Why do we have fontconfig 2.10?

worldofpeace commented:
Ah, I see it's backported to generate font caches
```
  # back-supported fontconfig version and package
  # version is used for font cache generation
  supportVersion = "210";
  supportPkg     = pkgs."fontconfig_${supportVersion}";

  # latest fontconfig version and package
  # version is used for configuration folder name, /etc/fonts/VERSION/
  # note: format differs from supportVersion and can not be used with makeCacheConf
  latestVersion  = pkgs.fontconfig.configVersion;
  latestPkg      = pkgs.fontconfig;
```

FRidh commented:
Debian unstable seems to have this beta version https://github.com/NixOS/nixpkgs/issues/73725#issuecomment-555663378.

worldofpeace commented:
I will circle back to resyncing this and seeing what we can do with compatibility. In the meantime anyone is free to re-PR the changes.

thongpv87 commented:
Hope this request will be merge soon. I got error that I can not start nix GUI app because of the fontconfig version on my machine is newer than nixpkgs fonconfig version

jonringer commented:
please resolve conflicts :)

NickHu commented:
I hope you don't mind my fixing the merge conflict (can squash out the merge commit when it comes time to merge into master).

I'd like to triage this because it seems like it's affecting lots of people, especially those who are using nix on non-NixOS systems; it looks like the only controversy here is that fontconfig has lots and lots of reverse-dependencies, but is there anything actually left to do here? As @worldofpeace said, we really need to update our fontconfig..

cc @worldofpeace @jtojnar @sauyon @FRidh @thongpv87 @jonringer

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-in-distress/3604/23

jtojnar commented:
@NickHu Number of reverse dependencies is not a problem at all, the stability and backwards compatibility is.

2.13.92 is an unstable version but if it is good enough for Arch, it should probably work for us too.

The worse issue is the compatibility so that people on stable NixOS can still use programs from unstable package set and vice versa.

Someone will need to figure out what does this upgrade mean for fontconfig cache and configuration, see https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-555810414 and https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-557705533.

Possibly, we might need to not fallback to `/etc/fontconfig` in https://github.com/NixOS/nixpkgs/blob/72066757c8bd31cc592914cb944f8d7da569d741/pkgs/development/libraries/fontconfig/config-compat.patch and always use versioned path. Similar action might need to be considered for cache.

NickHu commented:
Is the simplest solution just to backport this to stable too?

jtojnar commented:
I do not think backporting a major upgrade of core system component to stable is something we can do.

NickHu commented:
In line with this, if you look at https://repology.org/project/fontconfig/versions all the major distros seem to have moved on to at least 2.13.1 (latest stable, past the breaking changes AFAIK), including

* ubuntu 20.04 LTS (but also since 19.10)
* debian stable 😱
* Arch Linux
* Gentoo
* openSUSE
* Fedora 29 (current is 32)
* CentOS 7 (current is 8)

I think there's a decent case for suggesting that at least 2.13.1 is backported to stable

jtojnar commented:
@NickHu The point of stable is that it does not change under your feet. Backporting a major upgrade of a core system component would break that. The earliest release any fontconfig update will get into is 20.09 (if someone steps up to maintain it before then).

sauyon commented:
I think caches are versioned and therefore shouldn't be an issue, but configuration is.

I actually think fonts.conf should be part of the derivations, as they're not meant to be edited anyway. We'd need to patch it so that it reads from `/etc/fonts/conf.d` or maybe `/etc/fonts/version/conf.d` instead, but that should be easy enough.

For non-nixos, ideally there would be some way of specifying the system fontconfig version, which would then cause that version of fontconfig to read from the default `/etc/fonts/` directory.

I don't actually think there's a problem with a generated `fonts.conf` as long as it's contained within the derivation.

Issues may arise when font configurations require a certain version of fontconfig, and might require dumping multiple symlinks into etc for every `/version`, which doesn't seem very clean. Ideally those fonts would install only the config files for the current version of fontconfig, and only install the older config versions if those versions of fontconfig are installed, but that seems rather difficult to do in practice.

Was there any work done to upstream versioned configs?

jtojnar commented:
@sauyon

> I think caches are versioned and therefore shouldn't be an issue, but configuration is.

Hmm, it indeed looks like cache is versioned:

* https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/68d021b634d1bc53cf3ec1bd28a306500c68c23a/fc-cache/fc-cache.sgml#L195-205
* https://gitlab.freedesktop.org/fontconfig/fontconfig/blob/68d021b634d1bc53cf3ec1bd28a306500c68c23a/fontconfig/fontconfig.h#L60-70

> I actually think fonts.conf should be part of the derivations, as they're not meant to be edited anyway. We'd need to patch it so that it reads from `/etc/fonts/conf.d` or maybe `/etc/fonts/version/conf.d` instead, but that should be easy enough.

Sounds reasonable, we already configure fontconfig 2.12 to have `fonts.conf` pointing to `/etc/fonts/${version}/conf.d` so that would ensure the loaded configuration is always correct if it exists. Not sure what happens if the configuration does not exist there, maybe we will also need to add `<include>${fontconfig}/etc/fonts/conf.d</include>` to the `fonts.conf` but I am not sure if it is possible to only load it when the first `<include>` does not exist as to avoid clashes.

> For non-nixos, ideally there would be some way of specifying the system fontconfig version, which would then cause that version of fontconfig to read from the default `/etc/fonts/` directory.

I think ignoring the system config if there is not a `/etc/fonts/${version}/conf.d` should be sufficient. Especially if we fall back to the upstream config from the package.

> I don't actually think there's a problem with a generated `fonts.conf` as long as it's contained within the derivation.

We [only](https://github.com/NixOS/nixpkgs/blob/24f9bc0243ebf873cc0a34790dd0dd265d23f44f/nixos/modules/config/fonts/fontconfig.nix#L211-L212) copy it to the versioned directory so that is not issue either way. The issue is when non-derivation `fonts.conf` is picked up, which can only happen thanks to our patch:

https://github.com/NixOS/nixpkgs/blob/24f9bc0243ebf873cc0a34790dd0dd265d23f44f/pkgs/development/libraries/fontconfig/config-compat.patch#L25

I believe we chose that fallback to avoid loading random `fonts.conf` on `FONTCONFIG_PATH` and solution to all our problems should be as simple as removing the hunk from the patch and passing `-DFONTCONFIG_FILE=${placeholder "out"}/etc/fonts/fonts.conf"` to compile out:

https://gitlab.freedesktop.org/fontconfig/fontconfig/-/blob/68d021b634d1bc53cf3ec1bd28a306500c68c23a/src/fccfg.c#L2139-2141

Though, for robustness, we might want to replace it in the patch with `NIX_FONTCONFIG_FILE` constant so that it breaks loudly on change.

> Issues may arise when font configurations require a certain version of fontconfig, and might require dumping multiple symlinks into etc for every `/version`, which doesn't seem very clean. Ideally those fonts would install only the config files for the current version of fontconfig, and only install the older config versions if those versions of fontconfig are installed, but that seems rather difficult to do in practice.

The solution I propose here would mean the module only installs the latest `conf.d` and have other versions load the upstream defaults from the package.

But as I mentioned, I am not sure how will adding `<include>${fontconfig}/etc/fonts/conf.d</include>` behave for the latest package. If it really clashes as I suspect, we might need to patch fontconfig to load only one of the directories (adding `<includefirst>` element).

> Was there any work done to upstream versioned configs?

No idea.

jtojnar commented:
Looking at the module, we actually copy all files in the `conf.d` directory so there would be no conflicts:

https://github.com/NixOS/nixpkgs/blob/53eb2155e76eececd94a4a2f0dad03d1ad80c859/nixos/modules/config/fonts/fontconfig.nix#L217-L218

We only need to fix the routing of `local.conf` in the package itself:

https://github.com/NixOS/nixpkgs/blob/53eb2155e76eececd94a4a2f0dad03d1ad80c859/nixos/modules/config/fonts/fontconfig.nix#L220-L225

The only problematic part is the optional removal of the user config (hmm, so this is why it is not listed in `fonts.conf`):

https://github.com/NixOS/nixpkgs/blob/53eb2155e76eececd94a4a2f0dad03d1ad80c859/nixos/modules/config/fonts/fontconfig.nix#L236-L240

We would need to remove it in the package (or move it to `$out/etc/fonts/conf.defaults`) and import it in the module when needed. Of course this would disable user config on non-NixOS but we might say it is a good thing.

jtojnar committed:
fontconfig: Only read versioned config dirs

Falling back to unversioned `/etc/fonts/conf.d` when versioned one does not exist
is problematic since it only occurs on non-NixOS systems and those are likely
to have a different version of fontconfig. When those versions use incompatible
elements in the config, apps using fontconfig will crash.

Instead, we are now falling back to the in-package `fonts.conf` file that loads
both the versioned global `conf.d` directory and the in-package `conf.d` since using
upstream settings on non-NixOS is preferable to not being able to use apps there.

In fact, we would not even need to link `fonts.conf`, as the in-package `fonts.conf`
will be always used unless someone creates the global one manually (the option is still
retained if one wants to write a custom NixOS module and to avoid unnecessary stat call on NixOS).

Additionally, since the `fonts.conf` will always load `conf.d` from the package, we no longer
need to install them to sytem `/etc` in the module. This needed some mucking with `50-user.conf`
which disables configs in user directories (a good thing IMO, NixOS module will turn it back on)
but otherwise, it is cleaner. The files are still prioritized by their name, regardless of their location.

See https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-634370125 for more information.

jtojnar commented:
I pushed the proposed changes but it is a late hour here so if you could test this or double check my logic, it would be appreciated.

cc @vcunat

NickHu commented on pkgs/development/libraries/fontconfig/default.nix:
```diff
@@ -1,45 +1,78 @@
 /** Font configuration scheme
  - ./config-compat.patch makes fontconfig try the following root configs, in order:
-    $FONTCONFIG_FILE, /etc/fonts/${configVersion}/fonts.conf, /etc/fonts/fonts.conf
+    $FONTCONFIG_FILE, /etc/fonts/${configVersion}/fonts.conf, ${fontconfig}/etc/fonts/fonts.conf
```
I think this should be `${fontconfig.out}`

jtojnar replied:
Good catch, will update the comment.

NickHu reviewed (changes requested):
Looks good to me - would anyone who using nix on non-NixOS be able to test?

sauyon commented:
> The only problematic part is the optional removal of the user config (hmm, so this is why it is not listed in fonts.conf):

This explains a lot of my font configuration issues, hah.

> We would need to remove it in the package (or move it to `$out/etc/fonts/conf.defaults`) and import it in the module when needed. Of course this would disable user config on non-NixOS but we might say it is a good thing.

Perhaps versioned user configuration should also be a thing? Though I'm sure this won't be an issue for the vast majority of users.

Just to summarize and make sure I understand: configuration files will now only be searched for in and then `${fontconfig}/etc/fonts/conf.d`, which are symlinked into `/etc/fonts/${version}`, correct? Why do they need to be symlinked anyway?

I don't quite understand why the 2.10 fontconfig cache was ever necessary, and I feel like it shouldn't be necessary any longer; though perhaps the version should be bumped to 2.12? I feel like it would make sense at this point for the package to generate and maintain its own cache instead of doing it in the module.

jtojnar commented:
> Perhaps versioned user configuration should also be a thing? Though I'm sure this won't be an issue for the vast majority of users.

We could probablly do that but since we are ignoring user config outside of NixOS, it will probably be fine.

> Just to summarize and make sure I understand: configuration files will now only be searched for in and then `${fontconfig}/etc/fonts/conf.d`, which are symlinked into `/etc/fonts/${version}`, correct? Why do they need to be symlinked anyway?

The `fonts.conf` now includes `/etc/fonts/${version}/conf.d` (where the NixOS module puts it own overrides) and also `${fontconfig}/etc/fonts/conf.d`. IIRC fontconfig cares about the filename, not location so both directories will be processed at the same time.

We are no longer symlinking the `conf.d` to `/etc/fonts/${version}` since we include them from `fonts.conf`.

The main difference that we now always use the same `fonts.conf`, instead of falling back to unversioned global one that would include unversioned `conf.d`. 
> I don't quite understand why the 2.10 fontconfig cache was ever necessary, and I feel like it shouldn't be necessary any longer; though perhaps the version should be bumped to 2.12?

Maybe for Steam? No idea either, I would just scrap the “support” parts from the module but we can do that later so that we do not extend the scope of this PR even more.

> I feel like it would make sense at this point for the package to generate and maintain its own cache instead of doing it in the module.

IMHO the module for generating cache is useful since it allows us to rebuild the cache when system configuration changes. I prefer declarative methods like this to imperative hooks.

sauyon commented:
> We could probablly do that but since we are ignoring user config outside of NixOS, it will probably be fine.

This was so that user configurations could be supported on non-NixOS, though I think your point about not extending PR scope is a good reason not to do that here. It's probably only something to do if somebody complains about it anyway.

jtojnar commented:
I tried in Fedora VM and can confirm that FreeCAD no longer shows these warnings

```
[root@localhost-live liveuser]# /nix/store/91y2lsp822qb8cxprrkfnidm00bwn64c-freecad-0.18.4/bin/FreeCAD
FreeCAD 0.18, Libs: 0.18RUnknown
© Juergen Riegel, Werner Mayer, Yorik van Havre 2001-2019
  #####                 ####  ###   ####  
  #                    #      # #   #   # 
  #     ##  #### ####  #     #   #  #   # 
  ####  # # #  # #  #  #     #####  #   # 
  #     #   #### ####  #    #     # #   # 
  #     #   #    #     #    #     # #   #  ##  ##  ##
  #     #   #### ####   ### #     # ####   ##  ##  ##

Fontconfig warning: "/etc/fonts/fonts.conf", line 5: unknown element "its:rules"
Fontconfig warning: "/etc/fonts/fonts.conf", line 6: unknown element "its:translateRule"
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'translate'
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'selector'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'xmlns:its'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'version'
Fontconfig warning: "/etc/fonts/fonts.conf", line 9: unknown element "description"
Fontconfig error: Cannot load config file from /etc/fonts/fonts.conf
```

but it crashes:


```
[jan@localhost ~]$ /nix/store/qmplqyfkfjma2cm70nfwdgkvn42b67x5-freecad-0.18.4/bin/FreeCAD
FreeCAD 0.18, Libs: 0.18RUnknown
© Juergen Riegel, Werner Mayer, Yorik van Havre 2001-2019
  #####                 ####  ###   ####  
  #                    #      # #   #   # 
  #     ##  #### ####  #     #   #  #   # 
  ####  # # #  # #  #  #     #####  #   # 
  #     #   #### ####  #    #     # #   # 
  #     #   #    #     #    #     # #   #  ##  ##  ##
  #     #   #### ####   ### #     # ####   ##  ##  ##


(process:2863): Gtk-WARNING **: 23:29:03.105: Locale not supported by C library.
	Using the fallback 'C' locale.
Aborted (core dumped)
```

Do you have other programs that I could test?

jtojnar commented:
I confirmed it on `gedit`

```ShellSession
[jan@localhost ~]$ /nix/store/jcdvq7qgizfn403b3wa8m7i19j2gcx6j-gedit-3.36.2/bin/gedit 

(gedit:3327): Gtk-WARNING **: 00:12:37.163: Locale not supported by C library.
    Using the fallback 'C' locale.
```

compared to

```ShellSession
[jan@localhost ~]$ /nix/store/l2n8w3rzvyra37gpynrzy659h5aqvayl-gedit-3.36.2/bin/gedit 

(gedit:3160): Gtk-WARNING **: 23:53:37.191: Locale not supported by C library.
    Using the fallback 'C' locale.
Fontconfig warning: "/etc/fonts/fonts.conf", line 5: unknown element "its:rules"
Fontconfig warning: "/etc/fonts/fonts.conf", line 6: unknown element "its:translateRule"
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'translate'
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'selector'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'xmlns:its'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'version'
Fontconfig warning: "/etc/fonts/fonts.conf", line 9: unknown element "description"
Fontconfig error: Cannot load config file from /etc/fonts/fonts.conf
```

But in both cases the program just runs, regardless of the error. Seems like the issue was misattributed.



Now the only problematic workflow will be running programs compiled against fontconfig 2.12 on NixOS with 2.13, since that will not have 2.12 config dir and will fall back to the legacy fontconfig files. Maybe we should remove them altogether, or add an empty `fonts.conf` to prevent the fallthrough.

sauyon commented:
Tested this with Alacritty on Arch as well.

> Now the only problematic workflow will be running programs compiled against fontconfig 2.12 on NixOS with 2.13, since that will not have 2.12 config dir and will fall back to the legacy fontconfig files. Maybe we should remove them altogether, or add an empty `fonts.conf` to prevent the fallthrough.

I think the legacy config should be removed, though another alternative would be to backport these changes for 2.12 in stable.

jtojnar commented:
> I think the legacy config should be removed, though another alternative would be to backport these changes for 2.12 in stable.

Backporting complex fixes like this is out of question as mentioned above.


Fortunately, I tried the use case in question by applying the following patch on top of `nixos-unstable`:

```patch
--- a/nixos/modules/config/fonts/fontconfig.nix
+++ b/nixos/modules/config/fonts/fontconfig.nix
@@ -31,7 +31,7 @@ let
   # latest fontconfig version and package
   # version is used for configuration folder name, /etc/fonts/VERSION/
   # note: format differs from supportVersion and can not be used with makeCacheConf
-  latestVersion  = pkgs.fontconfig.configVersion;
+  latestVersion  = "2.13";
   latestPkg      = pkgs.fontconfig;
 
   # supported version fonts.conf
```

and it still seems to work with only a warning:

```
$ gedit
Fontconfig warning: "/etc/fonts/fonts.conf", line 146: blank doesn't take any effect anymore. please remove it from your fonts.conf
```

I would not worry about that much.

sauyon commented:
> Backporting complex fixes like this is out of question as mentioned above.

I would have advocated for just the config load order change, but it seems like the point is moot in any case, as I agree it doesn't seem like any changes are required.

sauyon commented:
I've discovered that `home-manager` has a fontconfig module that uses the user configuration. @rycee, do you have any thoughts on its removal?

rycee commented:
@sauyon Yeah HM does have a [fontconfig module](https://github.com/rycee/home-manager/blob/45abf3d38a2b51c00c347cab6950f3734e023bba/modules/misc/fontconfig.nix#LC1). It produces a fontconfig cache and a configuration that, if I recall correctly, is necessary for working user installed fonts on both NixOS and non-NixOS.

I install a bunch of fonts through the HM `home.packages` option (including some that includes additional configuration files in `…/etc/fonts/conf.d`) and the generated configuration on my system is
```xml
<fontconfig>
  <include ignore_missing="yes">/nix/store/303g5xci3s6h1yj9d0n2zp2x9b45fwb2-home-manager-path/etc/fonts/conf.d</include>
  <include ignore_missing="yes">/nix/store/303g5xci3s6h1yj9d0n2zp2x9b45fwb2-home-manager-path/etc/fonts/fonts.conf</include>

  <dir>/nix/store/303g5xci3s6h1yj9d0n2zp2x9b45fwb2-home-manager-path/lib/X11/fonts</dir>
  <dir>/nix/store/303g5xci3s6h1yj9d0n2zp2x9b45fwb2-home-manager-path/share/fonts</dir>
  <dir>/home/rycee/.nix-profile/lib/X11/fonts</dir>
  <dir>/home/rycee/.nix-profile/share/fonts</dir>

  <cachedir>/nix/store/303g5xci3s6h1yj9d0n2zp2x9b45fwb2-home-manager-path/lib/fontconfig/cache</cachedir>
</fontconfig>
```

If fontconfig 2.13 makes this configuration unnecessary then I'd be very happy. I tried to search for some kind of description of changes in fontconfig 2.13 but only found the commit logs, which is not very helpful. Does anybody happen to have a link to a higher level changelog?

jtojnar commented:
@rycee this is not a change in fontconfig but in our packaging. With 9396f5b8d6f276a2a93b31139598b692434cb656, we explicitly need to link `${fontconfig.out}/etc/fonts/conf.d.bak/50-user.conf` to `/etc/fonts/2.13/conf.d/50-user.conf` to be able to use user configuration.

The main change is that fontconfig will now only use configuration it is sure it is compatible with (or at least user needs to explicitly add it) and fall back to upstream default.

Of course if home-manager relies on `50-user.conf` that will not work. Either home-manager will need to set `FONTCONFIG_FILE` environment variable to point to its own `fonts.conf` (which side steps the system one, potentially without `50-user.conf`) or we need to choose a different approach (not sure what else would work except for giving up and hoping that it does not break too badly – when fontconfig cannot process the configs, it should still fall back to DejaVu).

rycee commented:
@jtojnar Ah, right. If the concern is that user provided fontconfig config files will be incompatible between versions then would it be possible to have some `50-versioned-user.conf` that is something like
```xml
<fontconfig>
  <include ignore_missing="yes" prefix="xdg">fontconfig/2.13/conf.d</include>
  <include ignore_missing="yes" prefix="xdg">fontconfig/2.13/fonts.conf</include>
</fontconfig
```
? I guess it would be a bit non-standard but perhaps it wouldn't break anything? Then HM could easily put the generated config there (as well).

NickHu commented:
@jtojnar I think there is a regression with the patches in this PR; now system-installed fonts are no longer properly found (they were before).

I've made a Docker image `nickhu/nix-fontconfig-update` to test (it's currently uploading, might take a while to be available). It's basically the archlinux base image, with nix (single-user mode) and gtk2fontsel installed three times:
* via Arch's repos
* via whatever you get from `nix-env -iA nixpkgs.gtk2fontsel` (i.e. nixpkgs aside from this PR)
* via `nix-env -f /path/to/this/checkout -iA gtk2fontsel` (i.e. nixpkgs at commit 9396f5b8d6f276a2a93b31139598b692434cb656 - your latest commit on this PR @jtojnar)
  + I've also pushed this build to cachix under `nickhu`

You can run X11 apps inside Docker via `nix-shell -p x11docker`.

Running the system's version of gtk2fontsel (`x11docker nickhu/nix-fontconfig-update gtk2fontsel`) produces no errors and gives:
![image](https://user-images.githubusercontent.com/450276/83166561-03ded680-a107-11ea-89a7-12dbcca5e24a.png)

Running the standard version in nixpkgs (`x11docker nickhu/nix-fontconfig-update /nix/store/wxhrj2cmzmag3rc43kr4b66g51q7zxl0-gtk2fontsel-0.1/bin/gtk2fontsel`) produces a slew of errors arising from fontconfig version incompatibility, but shows all the fonts:

```
(process:117): Gtk-WARNING **: 17:18:50.872: Locale not supported by C library.
	Using the fallback 'C' locale.
Fontconfig warning: "/etc/fonts/fonts.conf", line 5: unknown element "its:rules"
Fontconfig warning: "/etc/fonts/fonts.conf", line 6: unknown element "its:translateRule"
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'translate'
Fontconfig error: "/etc/fonts/fonts.conf", line 6: invalid attribute 'selector'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'xmlns:its'
Fontconfig error: "/etc/fonts/fonts.conf", line 7: invalid attribute 'version'
Fontconfig warning: "/etc/fonts/fonts.conf", line 9: unknown element "description"
Fontconfig error: Cannot load config file from /etc/fonts/fonts.conf
```

![image](https://user-images.githubusercontent.com/450276/83166802-55876100-a107-11ea-83b5-920ed514cd95.png)

Running the version based off this PR (`x11docker nickhu/nix-fontconfig-update /nix/store/s1h0hmd5i2bpfv7hglpkdvczpnza5bjk-gtk2fontsel-0.1/bin/gtk2fontsel`) produces an error about locales and gives:

```
(process:117): Gtk-WARNING **: 17:17:29.325: Locale not supported by C library.
	Using the fallback 'C' locale.
```

![image](https://user-images.githubusercontent.com/450276/83166710-37b9fc00-a107-11ea-946b-7aaf30071aca.png)

jtojnar commented:
@rycee

> @jtojnar Ah, right. If the concern is that user provided fontconfig config files will be incompatible between versions then would it be possible to include some `50-versioned-user.conf` in the package that contains something like

The issue with that is that there would be no way of disabling those includes, other than by rebuilding fontconfig. The only way forward I see is having home-manager set `FONTCONFIG_FILE` environment variable pointing to its own `fonts.conf`.

jtojnar commented:
@NickHu 

> I think there is a regression with the patches in this PR; now system-installed fonts are no longer properly found (they were before).

Then this works as intended.

We can either have a hermetic package as introduced here, or we need to accept that there might be errors when the host fontconfig configuration does not match the fontconfig used by the package. As always, there is no good answer when trading purity for functionality, and we need to choose the variant that better suits our goals.

sauyon commented:
> The only way forward I see is having home-manager set `FONTCONFIG_FILE` environment variable pointing to its own `fonts.conf`.

Is there a reason fontconfig isn't set up to load fonts installed into user profiles? That would solve this issue, right?

sauyon commented:
> @sauyon The `fonts.conf` file already contains `<dir prefix="xdg">fonts</dir>`, I guess the main issue is that the system fonts are not read.

But nix doesn't install fonts to `~/.local/share/fonts`, does it? I guess home-manager could link fonts into that directory by default, though I think it would make sense to have `<dir>~/.nix-profile/share/fonts</dir>` included by default for all systems.

jtojnar commented:
@sauyon That is already there, see the full listing:

```
<dir prefix="xdg">fonts</dir>
<dir>/nix/store/yyazdkikgnmg9rgdvbnrfn8k3q8wgf7v-dejavu-fonts-minimal-2.37</dir>
<dir>~/.nix-profile/lib/X11/fonts</dir>
<dir>~/.nix-profile/share/fonts</dir>
<dir>/nix/var/nix/profiles/default/lib/X11/fonts</dir>
<dir>/nix/var/nix/profiles/default/share/fonts</dir>
```

jtojnar commented:
@sauyon That is already there, see the full listing:

```
<dir prefix="xdg">fonts</dir>
<dir>/nix/store/yyazdkikgnmg9rgdvbnrfn8k3q8wgf7v-dejavu-fonts-minimal-2.37</dir>
<dir>~/.nix-profile/lib/X11/fonts</dir>
<dir>~/.nix-profile/share/fonts</dir>
<dir>/nix/var/nix/profiles/default/lib/X11/fonts</dir>
<dir>/nix/var/nix/profiles/default/share/fonts</dir>
```

rycee commented:
Don't forget `/etc/profiles/per-user/<username>/share/fonts`. Unfortunately, that one is trickier since it cannot, as far as I know, be represented statically in the fontconfig XML format.

jtojnar commented:
@rycee I do not think that one is necessary since on NixOS, the fonts will be managed by a NixOS module.

sauyon commented:
> @sauyon That is already there, see the full listing:

:facepalm: indeed, I'd assumed it wasn't because as it was before fontconfig never used its own `fonts.conf` (and therefore never included fonts installed by nix). I think I'd even be ok with not including system fonts for purity since now the ones installed via nix should be available.

rycee commented:
@jtojnar As far as I know, the NixOS module will only consider fonts that are installed in the system profile, not ones installed to per-user profiles through `users.users.<name?>.packages`. I might be wrong though, haven't checked that for a while.

dukc commented:
> Do you have other programs that I could test?

I have a similar problem, and can name a few cases that work badly probably because of this issue:

-Geany-1.36. It boots, but try to write underscores '_'. They show only with bold text, when the font size is 10 or more. Otherwise they are indistinguishable from whitespace.

-bless-0.6.2. The program boots, but every symbol in the main editing fields is replaced with hex value of the symbol the editor tries to show.

My `fonts.conf`:
```
<?xml version="1.0"?>
<!DOCTYPE fontconfig SYSTEM "fonts.dtd">
<!-- /etc/fonts/fonts.conf file to configure system font access -->
<fontconfig>
	<its:rules xmlns:its="http://www.w3.org/2005/11/its" version="1.0">
		<its:translateRule translate="no" selector="/fontconfig/*[not(self::description)]"/>
	</its:rules>

	<description>Default configuration file</description>

<!--
	DO NOT EDIT THIS FILE.
	IT WILL BE REPLACED WHEN FONTCONFIG IS UPDATED.
	LOCAL CHANGES BELONG IN 'local.conf'.

	The intent of this standard configuration file is to be adequate for
	most environments.  If you have a reasonably normal environment and
	have found problems with this configuration, they are probably
	things that others will also want fixed.  Please submit any
	problems to the fontconfig bugzilla system located at fontconfig.org

	Note that the normal 'make install' procedure for fontconfig is to
	replace any existing fonts.conf file with the new version.  Place
	any local customizations in local.conf which this file references.

	Keith Packard
-->

<!-- Font directory list -->

	<dir>/usr/share/fonts</dir>
	<dir>/usr/X11R6/lib/X11/fonts</dir> <dir>/opt/kde3/share/fonts</dir> <dir>/usr/local/share/fonts</dir>
	<dir prefix="xdg">fonts</dir>
	<!-- the following element will be removed in the future -->
	<dir>~/.fonts</dir>

<!--
  Accept deprecated 'mono' alias, replacing it with 'monospace'
-->
	<match target="pattern">
		<test qual="any" name="family">
			<string>mono</string>
		</test>
		<edit name="family" mode="assign" binding="same">
			<string>monospace</string>
		</edit>
	</match>

<!--
  Accept alternate 'sans serif' spelling, replacing it with 'sans-serif'
-->
	<match target="pattern">
		<test qual="any" name="family">
			<string>sans serif</string>
		</test>
		<edit name="family" mode="assign" binding="same">
			<string>sans-serif</string>
		</edit>
	</match>

<!--
  Accept deprecated 'sans' alias, replacing it with 'sans-serif'
-->
	<match target="pattern">
		<test qual="any" name="family">
			<string>sans</string>
		</test>
		<edit name="family" mode="assign" binding="same">
			<string>sans-serif</string>
		</edit>
	</match>

<!--
  Load local system customization file
-->
	<include ignore_missing="yes">conf.d</include>

<!-- Font cache directory list -->

	<cachedir>/var/cache/fontconfig</cachedir>
	<cachedir prefix="xdg">fontconfig</cachedir>
	<!-- the following element will be removed in the future -->
	<cachedir>~/.fontconfig</cachedir>

	<config>
<!--
  Rescan configuration every 30 seconds when FcFontSetList is called
 -->
		<rescan>
			<int>30</int>
		</rescan>
	</config>

</fontconfig>
```

NickHu commented:
Should we just merge this, and push any home-manager related compatibility tweaks to a further PR?

jtojnar committed:
fontconfig: Load fonts also from FHS paths

With previous patch, we no longer load non-versioned fonts.conf file to avoid incompatibilities
but this also means fontconfig will not load system-wide installed fonts on non-NixOS systems.

As a compromise, let's hardcode the FHS font paths to the built-in config so that the system
fonts work there. Unlike with the system config we do not need to worry about compatibility as
incompatible font files will be simply ignored.

Of course there will still be disparity if the system install fonts to some other location than
these two but I am afraid this is the best we can do.

See https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-635771967 for discussion.

jtojnar commented:
@rycee 

> @jtojnar As far as I know, the NixOS module will only consider fonts that are installed in the system profile, not ones installed to per-user profiles through `users.users.<name?>.packages`. I might be wrong though, haven't checked that for a while.

You are correct. At least this is not a regression so I will not address it here.



> @jtojnar Ah, right. If the concern is that user provided fontconfig config files will be incompatible between versions

Yeah, that was the primary concern for this changeset. But here the issue is that in order to keep `fonts.fontconfig.includeUserConf` NixOS option working, we needed to remove `50-user.conf` from the package.

This will not be an issue on NixOS, where the `fontconfig` module reintroduces it, but on other systems, packages using `fontconfig` from Nixpkgs will no longer load `fonts.conf` in the profile.

> then would it be possible to include some `50-versioned-user.conf` in the package that contains something like
>
> ```
> <fontconfig>
>   <include ignore_missing="yes" prefix="xdg">fontconfig/2.13/conf.d</include>
>   <include ignore_missing="yes" prefix="xdg">fontconfig/2.13/fonts.conf</include>
> </fontconfig>
> ```
> 
> ? I guess it would be a bit non-standard but perhaps it wouldn't break anything? Then HM could easily put the generated config there (as well).

We could do that but then we would not be able to disable it using `fonts.fontconfig.includeUserConf = false` as introduced in 1df1305a8a313120219e70ed1f50618b1047963d. Not sure if that would be much problem since the path would be only used by home-manager.

As an alternative, home-manager could set `FONTCONFIG_FILE` environment variable to point to its own `fonts.conf` as suggested above https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-635434294.

At this point I am unsure which one would be preferable.

----

@NickHu 

> Should we just merge this, and push any home-manager related compatibility tweaks to a further PR?

That is also an option that stakeholders would have to decide.

worldofpeace commented:
At this point home-manager should be in the NixOS org :grin:

worldofpeace commented:
@GrahamcOfBorg build fontcontig

jtojnar commented:
Replying to https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-555810414

> Does `configVersion` need to be changed,

Yes, config for 2.13 ([change introduced in 2.12.91](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/9a0fcb948fe7346f6c68028b2e54ab600a2a2a6f)) is not backwards compatible with 2.12. It is already bumped in the current revision.

> and what does this mean for `fontconfig_210`?

It will continue to use unprefixed config, though we might eventually want to remove that.

> Do we need three nested sets of conf directories now, or only the two as we have presently?

Only if we care for people using apps build against Nixpkgs’s fontconfig 2.12 on NixOS system with fontconfig 2.14.

We might not even need that if we backport https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/528b17b2837c3b102acd90cc7548d07bacaccb1f, assuming that the [warning](https://github.com/NixOS/nixpkgs/pull/73795#issuecomment-634952697) from unknown `description` does not cause the breakage. We could then even keep the old `configVersion`.

> (and how do we know/determine the answers to these questions?)

Need to run Arch/Manjaro VM, apply the [patch](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/commit/528b17b2837c3b102acd90cc7548d07bacaccb1f) to `/etc` and then run an app from current nixpkgs-unstable (fontconfig 2.12.6).

jtojnar committed:
fontconfig: remove its rules from configs

ITS rules are used for extracting translatable strings and they have
been moved to external files in 2.13.92 so they are not needed in
the config files themselves.

Removing them also cuts down on errors/warnings produced when using
older versions of fontconfig (< 2.12.92). Now it will only complain
about the description element but that is fortunately just a warning,
not errors like the ones caused by the its attributes.

Thanks to this, we can change the config version in NixOS module
back to 2.11 allowing us to re-use the 2.13/2.14 configs for apps
built against 2.12 fontconfig.

jtojnar commented:
What remains to decide if

1. we want to apply a patch like

```patch
--- a/pkgs/development/libraries/fontconfig/make-fonts-conf.xsl
+++ b/pkgs/development/libraries/fontconfig/make-fonts-conf.xsl
@@ -31,6 +31,9 @@
       <!-- versioned system-wide config -->
       <include ignore_missing="yes">/etc/fonts/<xsl:value-of select="$fontconfigConfigVersion" />/conf.d</include>
 
+      <!-- versioned per-user config -->
+      <include ignore_missing="yes" prefix="xdg">fontconfig/nixfc-<xsl:value-of select="$fontconfigConfigVersion" />-fonts.conf</include>
+
       <!-- upstream config -->
       <include><xsl:value-of select="$fontconfig" />/etc/fonts/conf.d</include>
 
```

and have home-manager install its config to `$XDG_CONFIG_HOME/fontconfig/nixfc-2.11-fonts.conf`, or

2. have home-manager set `FONTCONFIG_FILE` environment variable pointing to its config (possibly also have the config include `conf.d`(s) in `/etc`), or

3. do nothing for now and lose home-manager fontconfig support outside of NixOS.

cc @rycee

jtojnar commented:
The patch is incomplete, will also need to backport https://gitlab.freedesktop.org/fontconfig/fontconfig/-/merge_requests/100

flokli commented:
This should probably be re-evaluated with 2.10 being insecure - I assume we might want to end up using 2.13 as a default, having `$out/etc/fonts/conf.d` pointing to the 2.13 config, and maybe keeping config working for 2.11+2.12 in `$out/etc/fonts/2.11/conf.d`, or did I miss something here?

flokli commented:
Can we incorporate the removal of 2.10 (https://github.com/NixOS/nixpkgs/pull/92919) and its config from the module system in here?

jtojnar commented:
We only use configs from 2.10 so that it should be safe.

We cannot use non-versioned config location since it might break apps from the future and from the past.

I suggest we drop non-versioned config on NixOS altogether when we decide to stop supporting apps running fontconfig <2.11.

jtojnar commented:
But I would not do that in this pull request, it is already a big enough change and removing unversioned config dir would make it harder to determine what caused potential issue.

jtojnar commented:
Rebased.

flokli commented:
> But I would not do that in this pull request, it is already a big enough change and removing unversioned config dir would make it harder to determine what caused potential issue.

Okay. I assume such a change only is in the module system, so we could file the followuup PR to master once this has cycled through staging.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/thunderbird-crash-after-upgrade-to-20-09/9915/4

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-builds-a-ton-of-packages-instead-of-fetching-them-in-the-cache/13231/8
